File: ros2_ws/src/py_msg/py_msg/robot_status_publisher.py
from typing import Optional

import rclpy
from rclpy.node import Node
from rcl_interfaces.msg import Parameter, SetParametersResult
from tutorial_interfaces.msg import RobotStatus


class RobotStatusPublisher(Node):
    def __init__(self) -> None:
        super().__init__("robot_status_publisher")

        # 1. 声明参数
        self.declare_parameter("robot_name", "learning_bot")
        self.declare_parameter("initial_battery", 100.0)
        self.declare_parameter("publish_period", 1.0)

        # 2. 读取参数值
        self.robot_name: str = self.get_parameter("robot_name").value
        self.init_battery: float = self.get_parameter("initial_battery").value
        publish_period: float = self.get_parameter("publish_period").value

        self.add_on_set_parameters_callback(self.parameters_callback)

        self.publisher = self.create_publisher(RobotStatus, "robot_status", 10)
        self.timer = self.create_timer(publish_period, self.timer_callback)
        self.battery_level = self.init_battery

    def timer_callback(self) -> None:
        msg = RobotStatus()
        msg.robot_name = self.robot_name
        msg.battery_level = self.battery_level
        msg.is_active = True
        msg.joint_positions = [0.0, 0.5, -0.5, 1.0, -1.0]

        self.publisher.publish(msg=msg)
        self.get_logger().info(
            f"Publishing: robot_name={msg.robot_name}, "
            f"battery={msg.battery_level:.1f}, "
            f"joints={msg.joint_positions}"
        )

        self.battery_level -= 1.0
        if self.battery_level < 0.0:
            self.battery_level = self.init_battery

    def parameters_callback(self, params: list[Parameter]) -> SetParametersResult:
        for param in params:
            if param.name == "robot_name":
                self.robot_name = param.value
                self.get_logger().info(f"robot_name updated to: {param.value}")
        return SetParametersResult(successful=True)


def main(args: Optional[list[str]] = None) -> None:
    rclpy.init(args=args)
    node = RobotStatusPublisher()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__name__":
    main()

File: ros2_ws/src/cpp_msg/src/robot_status_publisher.cpp
#include "cpp_msg/robot_status_publisher.hpp"

#include <chrono>

namespace learning_ros2::cpp_msg {

RobotStatusPublisher::RobotStatusPublisher(const rclcpp::NodeOptions& options)
    : rclcpp::Node("robot_status_publisher", options),
      robot_name_("learning_bot"),
      battery_level_(100.0),
      initial_battery_(100.0),
      publish_period_(1.0) {
  this->declare_parameter("robot_name", robot_name_);
  this->declare_parameter("initial_battery", initial_battery_);
  this->declare_parameter("publish_period", publish_period_);

  robot_name_ = this->get_parameter("robot_name").as_string();
  initial_battery_ = this->get_parameter("initial_battery").as_double();
  battery_level_ = initial_battery_;
  publish_period_ = this->get_parameter("publish_period").as_double();

  publisher_ = this->create_publisher<RobotStatus>("robot_status", 10);

  auto period = std::chrono::duration<double>(publish_period_);
  timer_ = this->create_wall_timer(
      std::chrono::duration_cast<std::chrono::milliseconds>(period),
      std::bind(&RobotStatusPublisher::timer_callback, this));

  RCLCPP_INFO(
      this->get_logger(),
      "Configured: robot_name=%s, initial_battery=%.1f, publish_period=%.1f",
      robot_name_.c_str(), initial_battery_, publish_period_);
}

void RobotStatusPublisher::timer_callback() {
  RobotStatus msg;
  msg.robot_name = robot_name_;
  msg.battery_level = battery_level_;
  msg.is_active = true;
  msg.joint_positions = {0.0, 0.5, -0.5, 1.0, -1.0};

  publisher_->publish(msg);

  RCLCPP_INFO(this->get_logger(), "Publishing: robot_name=%s, battery=%.1f",
              msg.robot_name.c_str(), msg.battery_level);

  battery_level_ -= 1.0;
  if (battery_level_ < 0.0) {
    battery_level_ = initial_battery_;
  }
}

}  // namespace learning_ros2::cpp_msg
